### DISCUSSION TITLE
TypeState with Invoke

### DISCUSSION THREAD

**Discussion by codingedgar** [statelyai/xstate#1650] [Q&A]
Hi! I'm using typescript (4.1.2) with XState (4.14) I'm trying to use `createMahcine` with TypeStates, but all `invoke` (promises) configurations assume the `Context` is in the default version, inside the `invoke` I don't know how to do a `match` to see if the type narrowing kicks in, is there a right way to use TypeState with `invoke`? 

``` ts
type Context1 = {
  page?: puppeteer.Page,
  browser?: puppeteer.Browser, 
}
type Context2 = Context1 & {
  page: puppeteer.Page,
  browser: puppeteer.Browser,
}
const machine = createMachine<Context1, Event, State>({
  strict: true,
  states: {
    LAUNCHING: {},// <**** 1.
    SETTING_COOKIES: {
      invoke: {  // <***** 2.
        src: (ctx: Context2) => { }
     }
  }
})
```

1. I set 'Context1` to `Context2` here, tho not sure if it gets verified, does it gets verified a context is set before transitiones?
2. `invoke` expects `Context1`, no sure how to narrow `Context1` to `Context2` in `src`, is any way to let the type system know the context should be `Context2` here?

**Comment by Andarist** [comment]
AFAIK typestates only affect `.matches()` calls (and lines following it that can benefit from their results). So if you want to use your typestate within `invoke.src` you'd have to do smth like:
```js
    SETTING_COOKIES: {
      invoke: {  
        src: (_ctx) => {
          const ctx = _ctx as Context2
        }
     }
```

**Reply by ilovett** [reply]
Dang.  It seems like this should be easy to support to automatically infer the context in src...

**Reply by ilovett** [reply]
Could also shorten this a bit 

```
src: (context: Context2) => {
  ...
}
```

**Reply by Andarist** [reply]
> Dang. It seems like this should be easy to support to automatically infer the context in src...

This is far from easy :wink: 

> Could also shorten this a bit

This might not be possible because `context` won't be assignable as is to `Context2`
